lib/functions.php

```diff
@@ -12,9 +12,11 @@
      *
      * Use this function to create a coroutine-aware callable for a promise-aware callback caller.
      *
-     * @param callable(mixed ...$args): mixed $callback
+     * @template TValue
      *
-     * @return callable(mixed ...$args): \Amp\Promise
+     * @param callable():\Generator<mixed, mixed, mixed, TValue> $callback
```

Comment:
`$callback` is passed to `call()` which accepts many different types as return types from the callback, not just Generator.
See https://github.com/amphp/amp/blob/master/lib/functions.php#L65-L77

Reply:
I've updated to

```
     * @param callable(mixed ...$args): (\Generator<mixed, mixed, mixed, TValue>|\Amp\Promise<TValue>|mixed) $callback
```

If that makes sense.

Reply:
It would be better to use `TValue` instead of the last mixed but I'm guessing that runs into the same issue you mentioned above, right?

Reply:
Second mixed of the generator should possibly be `Promise<mixed>`. Afaik yielding anything else than a Promise would cause an error.

Reply:
Actually, that works. However not sure about the ReactPromise? Should we add that too?

```
     * @param callable(mixed ...$args): (\Generator<mixed, mixed, mixed, TValue>|\Amp\Promise<TValue>|\React\Promise\PromiseInterface<TValue>|TValue) $callback
```

Reply:
I guess... Personally I'm kinda against using React promises in Amphp code and would always call `adapt()` anyway but it works so I guess it's correct to use it.

By the way. technically the return value of the Generator is `Promise<TValue>|ReactPromise<TValue>|TValue`.

Reply:
cool, updated to use local names